{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create custom track to visualize depth of coverage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook we are going to:\n",
    "    - Create a BED file\n",
    "    - Use this BED file to extract the depth of coverage of a genome using *samtools bedcov*\n",
    "    - Use the output to create a wiggle custom track to visualize the depth of coverage in the UCSC browser\n",
    "    - Extract the depth of coverage from a second genome to be used as control\n",
    "    - Divide both coverages and create a wiggle custom track\n",
    "    - Identify a Copy Number Variant\n",
    "    - Find discordant reads and splitreads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we will import some libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will define the chromosome, and the start and end of our region of interest. For this exercise, we will use chr1:25,000,000-26,000,000. We also need to define the size of the blocks to count coverage (as number of bases per block). In our case, we will use 1,000 (we will count the total number of bases of reads mapped to those 1,000 bases intervals, from chr1:25,000,000 to chr1:26,000,000). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr = 5\n",
    "start = 180200000\n",
    "end = 180600000\n",
    "step = 1000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we will create a range from the start to the end, using step. This will create an array (*start_pos*) with all starting positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "start_pos = np.arange(start,end,step)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And we will iterate through every starting position in our array to print chromosome, starting position, and end position (starting position+1000). We will use the tab (*\"\\t\"*) separator, so each one of the three fields is separated by tabs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\t180200000\t180201000\n",
      "5\t180201000\t180202000\n",
      "5\t180202000\t180203000\n",
      "5\t180203000\t180204000\n",
      "5\t180204000\t180205000\n",
      "5\t180205000\t180206000\n",
      "5\t180206000\t180207000\n",
      "5\t180207000\t180208000\n",
      "5\t180208000\t180209000\n",
      "5\t180209000\t180210000\n",
      "5\t180210000\t180211000\n",
      "5\t180211000\t180212000\n",
      "5\t180212000\t180213000\n",
      "5\t180213000\t180214000\n",
      "5\t180214000\t180215000\n",
      "5\t180215000\t180216000\n",
      "5\t180216000\t180217000\n",
      "5\t180217000\t180218000\n",
      "5\t180218000\t180219000\n",
      "5\t180219000\t180220000\n",
      "5\t180220000\t180221000\n",
      "5\t180221000\t180222000\n",
      "5\t180222000\t180223000\n",
      "5\t180223000\t180224000\n",
      "5\t180224000\t180225000\n",
      "5\t180225000\t180226000\n",
      "5\t180226000\t180227000\n",
      "5\t180227000\t180228000\n",
      "5\t180228000\t180229000\n",
      "5\t180229000\t180230000\n",
      "5\t180230000\t180231000\n",
      "5\t180231000\t180232000\n",
      "5\t180232000\t180233000\n",
      "5\t180233000\t180234000\n",
      "5\t180234000\t180235000\n",
      "5\t180235000\t180236000\n",
      "5\t180236000\t180237000\n",
      "5\t180237000\t180238000\n",
      "5\t180238000\t180239000\n",
      "5\t180239000\t180240000\n",
      "5\t180240000\t180241000\n",
      "5\t180241000\t180242000\n",
      "5\t180242000\t180243000\n",
      "5\t180243000\t180244000\n",
      "5\t180244000\t180245000\n",
      "5\t180245000\t180246000\n",
      "5\t180246000\t180247000\n",
      "5\t180247000\t180248000\n",
      "5\t180248000\t180249000\n",
      "5\t180249000\t180250000\n",
      "5\t180250000\t180251000\n",
      "5\t180251000\t180252000\n",
      "5\t180252000\t180253000\n",
      "5\t180253000\t180254000\n",
      "5\t180254000\t180255000\n",
      "5\t180255000\t180256000\n",
      "5\t180256000\t180257000\n",
      "5\t180257000\t180258000\n",
      "5\t180258000\t180259000\n",
      "5\t180259000\t180260000\n",
      "5\t180260000\t180261000\n",
      "5\t180261000\t180262000\n",
      "5\t180262000\t180263000\n",
      "5\t180263000\t180264000\n",
      "5\t180264000\t180265000\n",
      "5\t180265000\t180266000\n",
      "5\t180266000\t180267000\n",
      "5\t180267000\t180268000\n",
      "5\t180268000\t180269000\n",
      "5\t180269000\t180270000\n",
      "5\t180270000\t180271000\n",
      "5\t180271000\t180272000\n",
      "5\t180272000\t180273000\n",
      "5\t180273000\t180274000\n",
      "5\t180274000\t180275000\n",
      "5\t180275000\t180276000\n",
      "5\t180276000\t180277000\n",
      "5\t180277000\t180278000\n",
      "5\t180278000\t180279000\n",
      "5\t180279000\t180280000\n",
      "5\t180280000\t180281000\n",
      "5\t180281000\t180282000\n",
      "5\t180282000\t180283000\n",
      "5\t180283000\t180284000\n",
      "5\t180284000\t180285000\n",
      "5\t180285000\t180286000\n",
      "5\t180286000\t180287000\n",
      "5\t180287000\t180288000\n",
      "5\t180288000\t180289000\n",
      "5\t180289000\t180290000\n",
      "5\t180290000\t180291000\n",
      "5\t180291000\t180292000\n",
      "5\t180292000\t180293000\n",
      "5\t180293000\t180294000\n",
      "5\t180294000\t180295000\n",
      "5\t180295000\t180296000\n",
      "5\t180296000\t180297000\n",
      "5\t180297000\t180298000\n",
      "5\t180298000\t180299000\n",
      "5\t180299000\t180300000\n",
      "5\t180300000\t180301000\n",
      "5\t180301000\t180302000\n",
      "5\t180302000\t180303000\n",
      "5\t180303000\t180304000\n",
      "5\t180304000\t180305000\n",
      "5\t180305000\t180306000\n",
      "5\t180306000\t180307000\n",
      "5\t180307000\t180308000\n",
      "5\t180308000\t180309000\n",
      "5\t180309000\t180310000\n",
      "5\t180310000\t180311000\n",
      "5\t180311000\t180312000\n",
      "5\t180312000\t180313000\n",
      "5\t180313000\t180314000\n",
      "5\t180314000\t180315000\n",
      "5\t180315000\t180316000\n",
      "5\t180316000\t180317000\n",
      "5\t180317000\t180318000\n",
      "5\t180318000\t180319000\n",
      "5\t180319000\t180320000\n",
      "5\t180320000\t180321000\n",
      "5\t180321000\t180322000\n",
      "5\t180322000\t180323000\n",
      "5\t180323000\t180324000\n",
      "5\t180324000\t180325000\n",
      "5\t180325000\t180326000\n",
      "5\t180326000\t180327000\n",
      "5\t180327000\t180328000\n",
      "5\t180328000\t180329000\n",
      "5\t180329000\t180330000\n",
      "5\t180330000\t180331000\n",
      "5\t180331000\t180332000\n",
      "5\t180332000\t180333000\n",
      "5\t180333000\t180334000\n",
      "5\t180334000\t180335000\n",
      "5\t180335000\t180336000\n",
      "5\t180336000\t180337000\n",
      "5\t180337000\t180338000\n",
      "5\t180338000\t180339000\n",
      "5\t180339000\t180340000\n",
      "5\t180340000\t180341000\n",
      "5\t180341000\t180342000\n",
      "5\t180342000\t180343000\n",
      "5\t180343000\t180344000\n",
      "5\t180344000\t180345000\n",
      "5\t180345000\t180346000\n",
      "5\t180346000\t180347000\n",
      "5\t180347000\t180348000\n",
      "5\t180348000\t180349000\n",
      "5\t180349000\t180350000\n",
      "5\t180350000\t180351000\n",
      "5\t180351000\t180352000\n",
      "5\t180352000\t180353000\n",
      "5\t180353000\t180354000\n",
      "5\t180354000\t180355000\n",
      "5\t180355000\t180356000\n",
      "5\t180356000\t180357000\n",
      "5\t180357000\t180358000\n",
      "5\t180358000\t180359000\n",
      "5\t180359000\t180360000\n",
      "5\t180360000\t180361000\n",
      "5\t180361000\t180362000\n",
      "5\t180362000\t180363000\n",
      "5\t180363000\t180364000\n",
      "5\t180364000\t180365000\n",
      "5\t180365000\t180366000\n",
      "5\t180366000\t180367000\n",
      "5\t180367000\t180368000\n",
      "5\t180368000\t180369000\n",
      "5\t180369000\t180370000\n",
      "5\t180370000\t180371000\n",
      "5\t180371000\t180372000\n",
      "5\t180372000\t180373000\n",
      "5\t180373000\t180374000\n",
      "5\t180374000\t180375000\n",
      "5\t180375000\t180376000\n",
      "5\t180376000\t180377000\n",
      "5\t180377000\t180378000\n",
      "5\t180378000\t180379000\n",
      "5\t180379000\t180380000\n",
      "5\t180380000\t180381000\n",
      "5\t180381000\t180382000\n",
      "5\t180382000\t180383000\n",
      "5\t180383000\t180384000\n",
      "5\t180384000\t180385000\n",
      "5\t180385000\t180386000\n",
      "5\t180386000\t180387000\n",
      "5\t180387000\t180388000\n",
      "5\t180388000\t180389000\n",
      "5\t180389000\t180390000\n",
      "5\t180390000\t180391000\n",
      "5\t180391000\t180392000\n",
      "5\t180392000\t180393000\n",
      "5\t180393000\t180394000\n",
      "5\t180394000\t180395000\n",
      "5\t180395000\t180396000\n",
      "5\t180396000\t180397000\n",
      "5\t180397000\t180398000\n",
      "5\t180398000\t180399000\n",
      "5\t180399000\t180400000\n",
      "5\t180400000\t180401000\n",
      "5\t180401000\t180402000\n",
      "5\t180402000\t180403000\n",
      "5\t180403000\t180404000\n",
      "5\t180404000\t180405000\n",
      "5\t180405000\t180406000\n",
      "5\t180406000\t180407000\n",
      "5\t180407000\t180408000\n",
      "5\t180408000\t180409000\n",
      "5\t180409000\t180410000\n",
      "5\t180410000\t180411000\n",
      "5\t180411000\t180412000\n",
      "5\t180412000\t180413000\n",
      "5\t180413000\t180414000\n",
      "5\t180414000\t180415000\n",
      "5\t180415000\t180416000\n",
      "5\t180416000\t180417000\n",
      "5\t180417000\t180418000\n",
      "5\t180418000\t180419000\n",
      "5\t180419000\t180420000\n",
      "5\t180420000\t180421000\n",
      "5\t180421000\t180422000\n",
      "5\t180422000\t180423000\n",
      "5\t180423000\t180424000\n",
      "5\t180424000\t180425000\n",
      "5\t180425000\t180426000\n",
      "5\t180426000\t180427000\n",
      "5\t180427000\t180428000\n",
      "5\t180428000\t180429000\n",
      "5\t180429000\t180430000\n",
      "5\t180430000\t180431000\n",
      "5\t180431000\t180432000\n",
      "5\t180432000\t180433000\n",
      "5\t180433000\t180434000\n",
      "5\t180434000\t180435000\n",
      "5\t180435000\t180436000\n",
      "5\t180436000\t180437000\n",
      "5\t180437000\t180438000\n",
      "5\t180438000\t180439000\n",
      "5\t180439000\t180440000\n",
      "5\t180440000\t180441000\n",
      "5\t180441000\t180442000\n",
      "5\t180442000\t180443000\n",
      "5\t180443000\t180444000\n",
      "5\t180444000\t180445000\n",
      "5\t180445000\t180446000\n",
      "5\t180446000\t180447000\n",
      "5\t180447000\t180448000\n",
      "5\t180448000\t180449000\n",
      "5\t180449000\t180450000\n",
      "5\t180450000\t180451000\n",
      "5\t180451000\t180452000\n",
      "5\t180452000\t180453000\n",
      "5\t180453000\t180454000\n",
      "5\t180454000\t180455000\n",
      "5\t180455000\t180456000\n",
      "5\t180456000\t180457000\n",
      "5\t180457000\t180458000\n",
      "5\t180458000\t180459000\n",
      "5\t180459000\t180460000\n",
      "5\t180460000\t180461000\n",
      "5\t180461000\t180462000\n",
      "5\t180462000\t180463000\n",
      "5\t180463000\t180464000\n",
      "5\t180464000\t180465000\n",
      "5\t180465000\t180466000\n",
      "5\t180466000\t180467000\n",
      "5\t180467000\t180468000\n",
      "5\t180468000\t180469000\n",
      "5\t180469000\t180470000\n",
      "5\t180470000\t180471000\n",
      "5\t180471000\t180472000\n",
      "5\t180472000\t180473000\n",
      "5\t180473000\t180474000\n",
      "5\t180474000\t180475000\n",
      "5\t180475000\t180476000\n",
      "5\t180476000\t180477000\n",
      "5\t180477000\t180478000\n",
      "5\t180478000\t180479000\n",
      "5\t180479000\t180480000\n",
      "5\t180480000\t180481000\n",
      "5\t180481000\t180482000\n",
      "5\t180482000\t180483000\n",
      "5\t180483000\t180484000\n",
      "5\t180484000\t180485000\n",
      "5\t180485000\t180486000\n",
      "5\t180486000\t180487000\n",
      "5\t180487000\t180488000\n",
      "5\t180488000\t180489000\n",
      "5\t180489000\t180490000\n",
      "5\t180490000\t180491000\n",
      "5\t180491000\t180492000\n",
      "5\t180492000\t180493000\n",
      "5\t180493000\t180494000\n",
      "5\t180494000\t180495000\n",
      "5\t180495000\t180496000\n",
      "5\t180496000\t180497000\n",
      "5\t180497000\t180498000\n",
      "5\t180498000\t180499000\n",
      "5\t180499000\t180500000\n",
      "5\t180500000\t180501000\n",
      "5\t180501000\t180502000\n",
      "5\t180502000\t180503000\n",
      "5\t180503000\t180504000\n",
      "5\t180504000\t180505000\n",
      "5\t180505000\t180506000\n",
      "5\t180506000\t180507000\n",
      "5\t180507000\t180508000\n",
      "5\t180508000\t180509000\n",
      "5\t180509000\t180510000\n",
      "5\t180510000\t180511000\n",
      "5\t180511000\t180512000\n",
      "5\t180512000\t180513000\n",
      "5\t180513000\t180514000\n",
      "5\t180514000\t180515000\n",
      "5\t180515000\t180516000\n",
      "5\t180516000\t180517000\n",
      "5\t180517000\t180518000\n",
      "5\t180518000\t180519000\n",
      "5\t180519000\t180520000\n",
      "5\t180520000\t180521000\n",
      "5\t180521000\t180522000\n",
      "5\t180522000\t180523000\n",
      "5\t180523000\t180524000\n",
      "5\t180524000\t180525000\n",
      "5\t180525000\t180526000\n",
      "5\t180526000\t180527000\n",
      "5\t180527000\t180528000\n",
      "5\t180528000\t180529000\n",
      "5\t180529000\t180530000\n",
      "5\t180530000\t180531000\n",
      "5\t180531000\t180532000\n",
      "5\t180532000\t180533000\n",
      "5\t180533000\t180534000\n",
      "5\t180534000\t180535000\n",
      "5\t180535000\t180536000\n",
      "5\t180536000\t180537000\n",
      "5\t180537000\t180538000\n",
      "5\t180538000\t180539000\n",
      "5\t180539000\t180540000\n",
      "5\t180540000\t180541000\n",
      "5\t180541000\t180542000\n",
      "5\t180542000\t180543000\n",
      "5\t180543000\t180544000\n",
      "5\t180544000\t180545000\n",
      "5\t180545000\t180546000\n",
      "5\t180546000\t180547000\n",
      "5\t180547000\t180548000\n",
      "5\t180548000\t180549000\n",
      "5\t180549000\t180550000\n",
      "5\t180550000\t180551000\n",
      "5\t180551000\t180552000\n",
      "5\t180552000\t180553000\n",
      "5\t180553000\t180554000\n",
      "5\t180554000\t180555000\n",
      "5\t180555000\t180556000\n",
      "5\t180556000\t180557000\n",
      "5\t180557000\t180558000\n",
      "5\t180558000\t180559000\n",
      "5\t180559000\t180560000\n",
      "5\t180560000\t180561000\n",
      "5\t180561000\t180562000\n",
      "5\t180562000\t180563000\n",
      "5\t180563000\t180564000\n",
      "5\t180564000\t180565000\n",
      "5\t180565000\t180566000\n",
      "5\t180566000\t180567000\n",
      "5\t180567000\t180568000\n",
      "5\t180568000\t180569000\n",
      "5\t180569000\t180570000\n",
      "5\t180570000\t180571000\n",
      "5\t180571000\t180572000\n",
      "5\t180572000\t180573000\n",
      "5\t180573000\t180574000\n",
      "5\t180574000\t180575000\n",
      "5\t180575000\t180576000\n",
      "5\t180576000\t180577000\n",
      "5\t180577000\t180578000\n",
      "5\t180578000\t180579000\n",
      "5\t180579000\t180580000\n",
      "5\t180580000\t180581000\n",
      "5\t180581000\t180582000\n",
      "5\t180582000\t180583000\n",
      "5\t180583000\t180584000\n",
      "5\t180584000\t180585000\n",
      "5\t180585000\t180586000\n",
      "5\t180586000\t180587000\n",
      "5\t180587000\t180588000\n",
      "5\t180588000\t180589000\n",
      "5\t180589000\t180590000\n",
      "5\t180590000\t180591000\n",
      "5\t180591000\t180592000\n",
      "5\t180592000\t180593000\n",
      "5\t180593000\t180594000\n",
      "5\t180594000\t180595000\n",
      "5\t180595000\t180596000\n",
      "5\t180596000\t180597000\n",
      "5\t180597000\t180598000\n",
      "5\t180598000\t180599000\n",
      "5\t180599000\t180600000\n"
     ]
    }
   ],
   "source": [
    "for i in start_pos:\n",
    "    print(chr,i,i+1000,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The problem with our output above is that it contains 1,000 lines, and we would need to copy it and paste it in our Linux terminal. Maybe it might be more convenient to export the results to a file. Let's redo the command including a few lines of code to save the output to a file. What does the cell below do?\n",
    "- First, it opens a file for writting, and it is assigned to an object call *f*.\n",
    "- The next two lines are equivalent to the previous cell (you iterate over each position, and for each starting position, you create a tuple containing the chromosoe number, the start position, and the end position).\n",
    "- The fourth line is the one that actually writes the output to a file:\n",
    "    - We use the command f.write to write, and inside the parenthesis what we want to write.\n",
    "    - We have a tuple (*res*) so we iterate over each element.\n",
    "    - We convert them to strings (str(x)).\n",
    "    - And we want fields to be separated by tabs, so we joined the elements using a tab (*'\\t'*)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open('my_positions_na.bed', 'w')\n",
    "for i in start_pos:\n",
    "    res = (chr,i,i+1000)\n",
    "    f.write('\\t'.join(str(x) for x in res)+'\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that you have your bed file (*my_positions.bed*), you can open you Linux terminal, and copy the bed file to your *home*. The path to write here is different depending on your system\n",
    "```console\n",
    "cp /mnt/c/Users/Your\\ User/jupyter_notebooks/my_positions.bed .\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And now you can use *samtools bedcov* to extract the bed coverage for a genome from the 1000genomes (HG00096) and save the results in HG00096.bedcov:\n",
    "```console\n",
    "    samtools bedcov my_positions.bed ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/phase3/data/HG00096/high_coverage_alignment/HG00096.wgs.ILLUMINA.bwa.GBR.high_cov_pcr_free.20140203.bam  > HG00096.bedcov\n",
    "```\n",
    "Extracting the coverage for 1 Mb in steps of 1 Kb using a remote BAM will take a while ($\\approx$40 min). Once done, you can copy the results back to your directory:\n",
    "```console\n",
    "cp HG00096.bedcov /mnt/c/Users/Your\\ User/jupyter_notebooks/\n",
    "```\n",
    "And you should have the file in your jupyter notebooks directory. We are going to import the data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "cov = pd.read_csv(\"HG00096.bedcov_na\",sep='\\t',header=None)\n",
    "cov_df = pd.DataFrame(cov)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can take a quick look at the results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>180200000</td>\n",
       "      <td>180201000</td>\n",
       "      <td>52356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>180201000</td>\n",
       "      <td>180202000</td>\n",
       "      <td>51276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>180202000</td>\n",
       "      <td>180203000</td>\n",
       "      <td>58793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>180203000</td>\n",
       "      <td>180204000</td>\n",
       "      <td>70377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>180204000</td>\n",
       "      <td>180205000</td>\n",
       "      <td>47026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>180205000</td>\n",
       "      <td>180206000</td>\n",
       "      <td>47567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5</td>\n",
       "      <td>180206000</td>\n",
       "      <td>180207000</td>\n",
       "      <td>63224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5</td>\n",
       "      <td>180207000</td>\n",
       "      <td>180208000</td>\n",
       "      <td>54735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>5</td>\n",
       "      <td>180208000</td>\n",
       "      <td>180209000</td>\n",
       "      <td>46491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5</td>\n",
       "      <td>180209000</td>\n",
       "      <td>180210000</td>\n",
       "      <td>48680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>5</td>\n",
       "      <td>180210000</td>\n",
       "      <td>180211000</td>\n",
       "      <td>53270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>5</td>\n",
       "      <td>180211000</td>\n",
       "      <td>180212000</td>\n",
       "      <td>55355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>5</td>\n",
       "      <td>180212000</td>\n",
       "      <td>180213000</td>\n",
       "      <td>53245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>5</td>\n",
       "      <td>180213000</td>\n",
       "      <td>180214000</td>\n",
       "      <td>37148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>5</td>\n",
       "      <td>180214000</td>\n",
       "      <td>180215000</td>\n",
       "      <td>43414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>5</td>\n",
       "      <td>180215000</td>\n",
       "      <td>180216000</td>\n",
       "      <td>39203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>5</td>\n",
       "      <td>180216000</td>\n",
       "      <td>180217000</td>\n",
       "      <td>42444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>5</td>\n",
       "      <td>180217000</td>\n",
       "      <td>180218000</td>\n",
       "      <td>37367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>5</td>\n",
       "      <td>180218000</td>\n",
       "      <td>180219000</td>\n",
       "      <td>48723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>5</td>\n",
       "      <td>180219000</td>\n",
       "      <td>180220000</td>\n",
       "      <td>33154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>5</td>\n",
       "      <td>180220000</td>\n",
       "      <td>180221000</td>\n",
       "      <td>44966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>5</td>\n",
       "      <td>180221000</td>\n",
       "      <td>180222000</td>\n",
       "      <td>53994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>5</td>\n",
       "      <td>180222000</td>\n",
       "      <td>180223000</td>\n",
       "      <td>65143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>5</td>\n",
       "      <td>180223000</td>\n",
       "      <td>180224000</td>\n",
       "      <td>44348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>5</td>\n",
       "      <td>180224000</td>\n",
       "      <td>180225000</td>\n",
       "      <td>37038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>5</td>\n",
       "      <td>180225000</td>\n",
       "      <td>180226000</td>\n",
       "      <td>57333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>5</td>\n",
       "      <td>180226000</td>\n",
       "      <td>180227000</td>\n",
       "      <td>45182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>5</td>\n",
       "      <td>180227000</td>\n",
       "      <td>180228000</td>\n",
       "      <td>61718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>5</td>\n",
       "      <td>180228000</td>\n",
       "      <td>180229000</td>\n",
       "      <td>53596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>5</td>\n",
       "      <td>180229000</td>\n",
       "      <td>180230000</td>\n",
       "      <td>38278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>5</td>\n",
       "      <td>180570000</td>\n",
       "      <td>180571000</td>\n",
       "      <td>52248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>5</td>\n",
       "      <td>180571000</td>\n",
       "      <td>180572000</td>\n",
       "      <td>70677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>5</td>\n",
       "      <td>180572000</td>\n",
       "      <td>180573000</td>\n",
       "      <td>69344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>5</td>\n",
       "      <td>180573000</td>\n",
       "      <td>180574000</td>\n",
       "      <td>55407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>5</td>\n",
       "      <td>180574000</td>\n",
       "      <td>180575000</td>\n",
       "      <td>43634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>5</td>\n",
       "      <td>180575000</td>\n",
       "      <td>180576000</td>\n",
       "      <td>40026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>5</td>\n",
       "      <td>180576000</td>\n",
       "      <td>180577000</td>\n",
       "      <td>48287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>5</td>\n",
       "      <td>180577000</td>\n",
       "      <td>180578000</td>\n",
       "      <td>43068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>5</td>\n",
       "      <td>180578000</td>\n",
       "      <td>180579000</td>\n",
       "      <td>34329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>5</td>\n",
       "      <td>180579000</td>\n",
       "      <td>180580000</td>\n",
       "      <td>43408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>5</td>\n",
       "      <td>180580000</td>\n",
       "      <td>180581000</td>\n",
       "      <td>53177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>5</td>\n",
       "      <td>180581000</td>\n",
       "      <td>180582000</td>\n",
       "      <td>39048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>5</td>\n",
       "      <td>180582000</td>\n",
       "      <td>180583000</td>\n",
       "      <td>43860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>5</td>\n",
       "      <td>180583000</td>\n",
       "      <td>180584000</td>\n",
       "      <td>51347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384</th>\n",
       "      <td>5</td>\n",
       "      <td>180584000</td>\n",
       "      <td>180585000</td>\n",
       "      <td>53322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>385</th>\n",
       "      <td>5</td>\n",
       "      <td>180585000</td>\n",
       "      <td>180586000</td>\n",
       "      <td>50294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386</th>\n",
       "      <td>5</td>\n",
       "      <td>180586000</td>\n",
       "      <td>180587000</td>\n",
       "      <td>35166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387</th>\n",
       "      <td>5</td>\n",
       "      <td>180587000</td>\n",
       "      <td>180588000</td>\n",
       "      <td>19840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388</th>\n",
       "      <td>5</td>\n",
       "      <td>180588000</td>\n",
       "      <td>180589000</td>\n",
       "      <td>31421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>5</td>\n",
       "      <td>180589000</td>\n",
       "      <td>180590000</td>\n",
       "      <td>37868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>5</td>\n",
       "      <td>180590000</td>\n",
       "      <td>180591000</td>\n",
       "      <td>41022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391</th>\n",
       "      <td>5</td>\n",
       "      <td>180591000</td>\n",
       "      <td>180592000</td>\n",
       "      <td>40073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>5</td>\n",
       "      <td>180592000</td>\n",
       "      <td>180593000</td>\n",
       "      <td>42602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>5</td>\n",
       "      <td>180593000</td>\n",
       "      <td>180594000</td>\n",
       "      <td>56567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>5</td>\n",
       "      <td>180594000</td>\n",
       "      <td>180595000</td>\n",
       "      <td>48893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>5</td>\n",
       "      <td>180595000</td>\n",
       "      <td>180596000</td>\n",
       "      <td>58570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>5</td>\n",
       "      <td>180596000</td>\n",
       "      <td>180597000</td>\n",
       "      <td>48627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>5</td>\n",
       "      <td>180597000</td>\n",
       "      <td>180598000</td>\n",
       "      <td>43058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>5</td>\n",
       "      <td>180598000</td>\n",
       "      <td>180599000</td>\n",
       "      <td>49790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>5</td>\n",
       "      <td>180599000</td>\n",
       "      <td>180600000</td>\n",
       "      <td>36713</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>400 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     0          1          2      3\n",
       "0    5  180200000  180201000  52356\n",
       "1    5  180201000  180202000  51276\n",
       "2    5  180202000  180203000  58793\n",
       "3    5  180203000  180204000  70377\n",
       "4    5  180204000  180205000  47026\n",
       "5    5  180205000  180206000  47567\n",
       "6    5  180206000  180207000  63224\n",
       "7    5  180207000  180208000  54735\n",
       "8    5  180208000  180209000  46491\n",
       "9    5  180209000  180210000  48680\n",
       "10   5  180210000  180211000  53270\n",
       "11   5  180211000  180212000  55355\n",
       "12   5  180212000  180213000  53245\n",
       "13   5  180213000  180214000  37148\n",
       "14   5  180214000  180215000  43414\n",
       "15   5  180215000  180216000  39203\n",
       "16   5  180216000  180217000  42444\n",
       "17   5  180217000  180218000  37367\n",
       "18   5  180218000  180219000  48723\n",
       "19   5  180219000  180220000  33154\n",
       "20   5  180220000  180221000  44966\n",
       "21   5  180221000  180222000  53994\n",
       "22   5  180222000  180223000  65143\n",
       "23   5  180223000  180224000  44348\n",
       "24   5  180224000  180225000  37038\n",
       "25   5  180225000  180226000  57333\n",
       "26   5  180226000  180227000  45182\n",
       "27   5  180227000  180228000  61718\n",
       "28   5  180228000  180229000  53596\n",
       "29   5  180229000  180230000  38278\n",
       "..  ..        ...        ...    ...\n",
       "370  5  180570000  180571000  52248\n",
       "371  5  180571000  180572000  70677\n",
       "372  5  180572000  180573000  69344\n",
       "373  5  180573000  180574000  55407\n",
       "374  5  180574000  180575000  43634\n",
       "375  5  180575000  180576000  40026\n",
       "376  5  180576000  180577000  48287\n",
       "377  5  180577000  180578000  43068\n",
       "378  5  180578000  180579000  34329\n",
       "379  5  180579000  180580000  43408\n",
       "380  5  180580000  180581000  53177\n",
       "381  5  180581000  180582000  39048\n",
       "382  5  180582000  180583000  43860\n",
       "383  5  180583000  180584000  51347\n",
       "384  5  180584000  180585000  53322\n",
       "385  5  180585000  180586000  50294\n",
       "386  5  180586000  180587000  35166\n",
       "387  5  180587000  180588000  19840\n",
       "388  5  180588000  180589000  31421\n",
       "389  5  180589000  180590000  37868\n",
       "390  5  180590000  180591000  41022\n",
       "391  5  180591000  180592000  40073\n",
       "392  5  180592000  180593000  42602\n",
       "393  5  180593000  180594000  56567\n",
       "394  5  180594000  180595000  48893\n",
       "395  5  180595000  180596000  58570\n",
       "396  5  180596000  180597000  48627\n",
       "397  5  180597000  180598000  43058\n",
       "398  5  180598000  180599000  49790\n",
       "399  5  180599000  180600000  36713\n",
       "\n",
       "[400 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You see that there are 1,000 rows and 4 columns, as we were expecting. The columns are not labeled, but they are chromosome, start and end positions, and the 4<sup>th</sup> column is the number of bases in that region. You can see that there is some variability between each 1,000 region. We can explore the file using the *describe* function for a pandas dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>400.0</td>\n",
       "      <td>4.000000e+02</td>\n",
       "      <td>4.000000e+02</td>\n",
       "      <td>400.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>5.0</td>\n",
       "      <td>1.803995e+08</td>\n",
       "      <td>1.804005e+08</td>\n",
       "      <td>47279.962500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.156143e+05</td>\n",
       "      <td>1.156143e+05</td>\n",
       "      <td>13105.904535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>5.0</td>\n",
       "      <td>1.802000e+08</td>\n",
       "      <td>1.802010e+08</td>\n",
       "      <td>10504.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>5.0</td>\n",
       "      <td>1.802998e+08</td>\n",
       "      <td>1.803008e+08</td>\n",
       "      <td>39106.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>5.0</td>\n",
       "      <td>1.803995e+08</td>\n",
       "      <td>1.804005e+08</td>\n",
       "      <td>47958.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>5.0</td>\n",
       "      <td>1.804992e+08</td>\n",
       "      <td>1.805002e+08</td>\n",
       "      <td>55328.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>5.0</td>\n",
       "      <td>1.805990e+08</td>\n",
       "      <td>1.806000e+08</td>\n",
       "      <td>109563.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0             1             2              3\n",
       "count  400.0  4.000000e+02  4.000000e+02     400.000000\n",
       "mean     5.0  1.803995e+08  1.804005e+08   47279.962500\n",
       "std      0.0  1.156143e+05  1.156143e+05   13105.904535\n",
       "min      5.0  1.802000e+08  1.802010e+08   10504.000000\n",
       "25%      5.0  1.802998e+08  1.803008e+08   39106.500000\n",
       "50%      5.0  1.803995e+08  1.804005e+08   47958.500000\n",
       "75%      5.0  1.804992e+08  1.805002e+08   55328.000000\n",
       "max      5.0  1.805990e+08  1.806000e+08  109563.000000"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we just look at the 4<sup>th</sup> column, we can see that the mean is 45,868 bases, but there is a minimum of 13,620 and a maximum of 77,959. That means that, if we have near 46,000 bases of sequence per each 1,000 base interval, our depth of coverage is 46$\\times$. That coverage is higher than usual for a whole genome, but is the one used for the 1000genomes project high coverage. What we are going to do now, is to create a *wiggle* track to visualize the coverage in the UCSC genome browser. In this wiggle track, we will use:\n",
    "- The *fixedStep* format, in which we define in which chromosome and position do we start, and which is our step (in our case was 1,000 bases).\n",
    "- Then we just put the values (the number of bases), and the track knows that each one of them is 1,000 bases apart from the previous one.\n",
    "- We are going to define also the span to 1,000 bases. The wiggle will create a histogram, and with the span we are telling the track that we want the base to be 1,000 bases.\n",
    "- We will export everything to a file called HG00096.wig that we will later upload to the UCSC genome browser."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open('HG00096_na.wig', 'w')\n",
    "f.write('browser position chr5:180200000-180500000' + '\\n')\n",
    "f.write('track type=wiggle_0 name=\"fixedStep\" description=\"fixedStep format\" visibility=full autoScale=off viewLimits=0.0:109000.0 color=50,150,255 yLineMark=47000 yLineOnOff=on priority=10' + '\\n')\n",
    "f.write('fixedStep chrom=chr5 start=180200000 step=1000 span=1000' + '\\n')\n",
    "for index, row in cov_df.iterrows():\n",
    "    f.write(cov.loc[index,3].astype(str) + '\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we are ready to visualize the coverage of this genome in this region of chromosome 1. We'll go to the UCSC genome browser (http://genome.ucsc.edu) and:\n",
    "- select **human assembly GRCh37/hg19** (because this is the version that the 1000genomes mapped the HG00096 reads to) and click **Go**.\n",
    "- Right below the main window, you'll find a button saying \"**add Custom tracks**\", click on it.\n",
    "- To upload the track we have just created, click on **Browse...** and select the file from your computer (*HG00096.wig*).\n",
    "- Click **Submit**.\n",
    "- You should get a track called *fixedStep*, type *wiggle_0*. To go an view it click on **go**.\n",
    "You should now see chr1:25,000,000-26,000,000 with the track on blue on the top, such as in the image below:\n",
    "![title](05_coverage_01v3.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first thing you notice is that the coverage is very noisy. This is due to several reasons, including the GC content (regions with a high GC content are more difficult to PCR-amplify, so you usually have lower coverage), to the presence of repetitive regions,...\n",
    "In order to smooth our track, we are going to click on the **grey bar on the left of the track**, and a menu with several options will appear. We are going to perform the following modifications:\n",
    "- We are going to change the **smoothing window** to 16.\n",
    "- As we know that the average number of bases is 45,868, we are going to **Draw y indicator lines:** at **y=45868**.\n",
    "- Click **Submit**.\n",
    "Now, we should have a more smoothed coverage track, as that in the figure below:\n",
    "![title](05_coverage_02v3.png)\n",
    "You can see that the coverage track is now smoothed, and most of this Megabase of sequence has the same coverage, with the exception of a small region around 25,600,000, in which we can see a drop in coverage to about half of the rest of this region. Therefore, it looks like there is a deletion affecting this region. If you take a look at the track just below the coverage, you can see that this deletion only affects one gene, named *RHD*. Click on that gene to see what it codes for.\n",
    "Do you recognize this gene? Of course, it encodes *Homo sapiens* Rh blood group, D antigen, responsible for the Rh blood group. As you know, people are either Rh+ or Rh-, depending on whether they have this antigen or not. Rh- persons do not have a functional copy of the *RHD* gene, while Rh+ persons have it. As we are diploid, we have 2 copies of each chromosome (and usually, of all genes within that chromosome). In the case of the sample we are analyzing (HG00096), it appears that this person is diploid for most of the region we are exploring, with the exception of *locus RHD*, which has half the coverage than the other genes. This means that instead of having two copies of this gene, it only has one. Therefore, this person is Rh+, as at least has one copy of the gene, but is heterozygous for this deletion."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This *locus* allows us to try to identify the breakpoints, because one of the alleles should present a big deletion involving the *RHD* gene. First, we will zoom in into this region to more precisely see where the deletion starts and ends. You can see a zoomed image as that shown below.\n",
    "![title](05_coverage_03v3.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The coverage is still noisy. We can see that the drop in coverage goes between 25,590,000 and 25,650,00, although due to the noise, those limits are not very precise. According to the 1000genomes, this CNV goes from 25,592,642 to 25,661,222, very close to the limits we obtained by visual inspection. We could try to improve the resolution and even try to identify the precise breakpoints by using the mate-pair information and the CIGAR. Unfortunately, this region is very repetitive, and together with the low coverage of 1000genomes sequences, this region is not suitable to this kind of analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Luckily, there are other CNVs we can take advantage of to learn how to process this information. One such CNV is a deletion in chr1:72,766,343-72,811,815 (45 Kb). We will try to identify the breakpoints with higher accuracy by analyzing the BAM file, and more specifically, the mate-pair information.\n",
    "\n",
    "To do that:\n",
    "- First, we will view the BAM file for that region using *samtools view*, and we will use *awk* to select only those reads with an apparent insert size larger than 1,000. As the fragments insert sizes are about 500 bp, we will view the region extending 1,000 bases to each side:\n",
    "```console\n",
    "samtools view ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/phase3/data/HG00096/high_coverage_alignment/HG00096.wgs.ILLUMINA.bwa.GBR.high_cov_pcr_free.20140203.bam 1:72765343-72812815|awk '$9>1000' > HG00096.chr1_72.sam\n",
    "```\n",
    "- You can explore the output file by using *less HG00096.chr1_72.sam* (**NOTE**: To quit *less*, just type 'q').\n",
    "- The problem is that the output is too dense. So, in order to simplify the output, we will use *awk* to just print the most informative columns. In your terminal \n",
    "```console\n",
    "awk '{print $3\"\\t\"$4\"\\t\"$5\"\\t\"$6\"\\t\"$7\"\\t\"$8\"\\t\"$9}' HG00096.chr1_72.sam\n",
    "```\n",
    "- Now, you should get a result as the one shown below:\n",
    "```console\n",
    "1       72765712        60      250M        =       203693994       130928076\n",
    "1       72765819        60      250M        =       72811840        46237\n",
    "1       72765987        60      250M        =       72811840        46097\n",
    "1       72765993        59      184M66S     =       72811994        46251\n",
    "1       72766001        60      250M        =       72811860        46109\n",
    "1       72766008        60      250M        =       72811840        46054\n",
    "1       72766091        60      234M16S     =       72811840        45886\n",
    "1       72766092        60      233M17S     =       72811840        45886\n",
    "1       72766120        60      205M45S     =       72811838        45968\n",
    "1       72766124        60      201M49S     =       72811955        46081\n",
    "1       72766137        60      188M62S     =       72811840        45894\n",
    "1       72766150        60      175M75S     =       72811838        45938\n",
    "1       72766150        60      157M93S     =       158166826       85400521\n",
    "1       72766151        60      174M76S     =       72811918        46017\n",
    "1       72766156        60      169M81S     =       72811840        45916\n",
    "1       72766180        60      145M105S    =       72811888        45958\n",
    "1       72766203        60      122M128H    =       72811855        45902\n",
    "1       72766209        60      116M134H    =       72811931        45972\n",
    "1       72766234        60      87M163H     =       72812283        46299\n",
    "1       72766238        60      87M163H     =       72812214        46226\n",
    "1       72766240        60      85M165H     =       72811890        45900\n",
    "1       72766261        60      64M186H     =       72812041        46030\n",
    "1       72766288        60      37M213H     =       72812068        46030\n",
    "1       72768731        60      250M        =       193831512       121062829\n",
    "1       72784230        0       5S63M182S   =       91505639        18721410\n",
    "1       72786789        60      51S77M122S  =       168739245       95952457\n",
    "1       72793697        0       59S25M166S  =       223453796       150660325\n",
    "1       72803303        60      63S187M     =       142841358       70038060\n",
    "1       72803545        0       90H40M120H  =       226526069       153722486\n",
    "1       72805742        60      250M        =       106203355       33397365\n",
    "1       72811860        56      194H56M     =       158166826       85354912\n",
    "```\n",
    "\n",
    "- On the right-most column we can see the apparent insert size. You can see that there are different sizes, but there is one size that appears to be very frequent. From the 2$^{nd}$ to the 23$^{rd}$ lines, all of them but one have an apparent insert size of about 46,000 bases.\n",
    "- If we take a look at the second line, the read maps at chromosome 1, position 72,765,819, while its mate maps at 72,811,840 (that correspond more or less where the deletion should be).\n",
    "- All the other lines with an apparent insert size of 46,000, map nearby, and the mate also nearby to the mate of the second read, suggesting that we have many independent fragments from the DNA library that happen to support this event.\n",
    "- If we have fragments of 500 bp, and a fragment contains the breakpoint, the reads could not be more than 500 bases appart from the breakpoint. Thus, all independent reads supporting the deletion should map very close together. In our case, the first of those reads maps at position 72,765,819, while the last one at position 72,766,288, that means that they are 419 bases apart. Similarly, the mates map between 728,11,838 and 72,812,283, also 445 bases apart.\n",
    "- By contrast, the other reads with apparent insert sizes larger than 1 Kb, are all very large, and there is only 1 read supporting each potential SV, suggesting that all of them constitute false positives.\n",
    "- Finally, you can see that the mapping quality for all reads supporting the 46,000 deletion is very high (60), while for some of the reads we suspected they were false positives, the mapping quality is 0, meaning that our confidence that the read is properly mapped (and therefore supporting the presence of an SV) is very low, reafirming our suspicion of those reads being false positives.\n",
    "\n",
    "In summary, we have 21 reads supporting the presence of a large deletion of about 46,000 bases (range 45,886-46,299). We can estimate that the breakpoint on the left is close to 72,766,288 (the right-most read), and the breakpoint on the left is close to 72,811,838."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although the use of NGS has allowed us to narrow the positions of the breakpoints, we can try to be even more precise, defining the breakpoints at base-pair resolution. Considering the high coverage we have, and the fact that the reads lengths used are very long (250 bases), it is very likely that there are a couple of reads that cover the breakpoint. In this case, part of the read will belong to the left part of the breakpoint, while the rest of the read will belong to the sequence at the right-hand side (that is, 46,000 bases to the right). In this case, the alignment programs are not able to properly align the whole read. If part of the read aligns pretty well to a region, scanning the rest of the genome to try to find where the rest of the read aligns is very computing intensive, so it not usually done to keep the performance of the algorithm. In those cases, the read is align where there are more bases aligning, and then, it performs soft-clipping (the alignment will refer to the part that aligns, but not to the rest of the read). Those reads are known as splitreads. And those are precisely the ones we are interested in.\n",
    "\n",
    "If we just take a look at the CIGAR, to see if any of those reads had soft-clipping, we can see that many of them are soft-clipped. For the sake of simplicity, we are going to show just one, the one with the largest fargment soft-clipped:\n",
    "```console\n",
    "1       72766180        60      145M105S    =       72811888        45958\n",
    "```\n",
    "Now, we will go to the *HG00096.chr1_72.sam* file to find this read:\n",
    "```console\n",
    "H7AGFADXX131213:2:1201:9933:68924       161     1       72766180        60      145M105S        =       72811888        45958   TGGATAAATATGTGAGGGAGACAATATAAAACAAAAGTTCACATGGAACATACATTAATAAAATACTGCAGAAGCAAAAAAAAGTGAATCATTCAAAAATGAAAAGGGTACAAGACAAAAATAAAAGTTAAAGAAAAACCAAATGACTGAGAGATTATCTCAGTGGATTTGACTTAATCTCTTGAACCACTTTCTAAAAACAGTTTTATTTGGCTAGTTGCAGAAAAGAAAATCAGAGATTCAAAGCTTT ;::>:>>@<>>>7=?>==?8><:<;<>>>>><????>>?>><>>><@?<>:;:><????>@@@>>=<<:9>?@>>@@@@@@@A@/:@??;A@A?AAAA@>@@AAA>=<??=@@?@>@B=?@??@@@??@?A@?@A@AA<@A?@>>A>@==>=8A?A@@@?=?>:9??=?@@@=A@@A?<>>A8>>?>=B?AA>@A?@@@?&?;@?AA@@8><>@<;?@+<9?>9>AA@BAA?>6@.+9+>A@AAA##### SA:Z:1,72811840,+,144S106M,60,3;   BD:Z:IIJIOIMFJGHKIIMLMIILKLHJIFGFJDDJHJDDKMJILHHIJJIIJHIFKHIHJJIFJDDIFKKMNNNKIKONJDDDDDDKMIMIILLIHILJDDDIJMIDDKMILKKHJKKLHJDDDIFJDDKMJJJDKKIDDDJMIJDIJMLKMMLKLKKHJGLJJLNMIJIKHCKMLKJJJILJJJJKMIJMIHKJCIJKJDDDJHNMJCCJGIDLKOOLNNKLOOOLJEELLJEEKNNPMNMNKLPNINIIII    RG:Z:H7AGF.2    BI:Z:KKKLPJNHLHILLKOMNLKLLLJMJHIHLGFLKMGGNOLKNJKLLKLKLKLHMKLKMLKHLGGKHMNPOPOMKNRPMGGGGGGNOKOKKNNLKKNMGGGKLOKGGNNLOMMKMNMMKMGGGKHLGGNOLMLGNMKGGGLNLMGKLPNNPPNNNNOLNJONMOPPLLMOLGNPNOMNMLONMNMNPLMOMKOMGLNPMHHHMLPPMHGOKMHOMRSQPQNOQRQOMIIPONJJNQQRPQQRPPSRLQKKKK    NM:i:0  BQ:Z:@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@    AS:i:145   XS:i:23\n",
    "```\n",
    "If you look carefully that information that the aligner provides, you can see that, right after the base quality information, there is one field with this information: *SA:Z:1,72811840,+,144S106M,60,3;*. The *SA* refers to secondary alignment, and the position where this read might also map is chr1:72,811,840. Right on the other breakpoint! The CIGAR is also provided (144S106M), meaning that if we map this read to this other locus, we need to soft-clip the first 144 bases of the read, and then the remaining 106 can be aligned. This more or less the opposite of what we have as primary alignment (CIGAR: 145M105S), in which the first 145 bases align, and the remaining 105 are soft-clipped.In addition we also have the mapping quality for the secondary alignment (60), so the confidence that, if we soft-clip the first bases we can get a good alignment, is high.\n",
    "\n",
    "But this is what our aligner (BWA) says. Let's get a second opinion by using BLAT. If we just copy the sequence and perform a BLAT (remember to use GRCh37/hg19 to get the same coordinates than our genome), we should get the following results:\n",
    "```console\n",
    "  ACTIONS      QUERY   SCORE START   END QSIZE IDENTITY  CHROM  STRAND  START       END   SPAN\n",
    "-----------------------------------------------------------------------------------------------\n",
    "browser details YourSeq   243     1   250   250    98.8%  chr1   +    72766180  72811945  45766\n",
    "browser details YourSeq    42   118   213   250    95.7%  chr3   +   186040363 186201121 160759\n",
    "browser details YourSeq    23    99   124   250    96.0%  chr5   -    77769099  77769131     33\n",
    "browser details YourSeq    23   117   143   250    88.0%  chr4   -    46220860  46220885     26\n",
    "browser details YourSeq    23   126   148   250   100.0%  chr4   -    11167930  11167952     23\n",
    "browser details YourSeq    23   186   208   250   100.0%  chr11  -    92916686  92916708     23\n",
    "browser details YourSeq    22    21    44   250    87.0%  chr3   -      490524    490546     23\n",
    "browser details YourSeq    22    85   107   250   100.0%  chr2   -   103296303 103296332     30\n",
    "browser details YourSeq    22   219   240   250   100.0%  chr17  -    17370961  17370982     22\n",
    "browser details YourSeq    21   118   138   250   100.0%  chr5   -   154514088 154514108     21\n",
    "browser details YourSeq    21   118   138   250   100.0%  chr2   -   120875925 120875945     21\n",
    "browser details YourSeq    21   118   138   250   100.0%  chr4   +    41765684  41765704     21\n",
    "browser details YourSeq    21   118   138   250   100.0%  chr3   +   151746469 151746489     21\n",
    "browser details YourSeq    20   119   138   250   100.0%  chr5   -    25550364  25550383     20\n",
    "browser details YourSeq    20   119   138   250   100.0%  chr5   -    17836288  17836307     20\n",
    "browser details YourSeq    20    28    49   250    95.5%  chr2   -   211122061 211122082     22\n",
    "browser details YourSeq    20    28    49   250    95.5%  chr3   +   137116000 137116021     22\n",
    "browser details YourSeq    20   118   137   250   100.0%  chr3   +    69691667  69691686     20\n",
    "```\n",
    "\n",
    "You can see that the first hit is the only one that is able to align the total length of the read (250 bases), and it aligns with the 98.8% identity (the remaining results only align part of the read, as little as 20 bases, so they don't look as credible results). However, if you take a look at the span, you can see that it does not span 250 bases (as it would be expected for a 250 bases read), but it spans 45,766 bases, which is exactly what you would expect if there was a deletion of $$45766-250 = 45516~bases$$ That means that this read lies right at the breakpoint, and the breapoint has been sequenced (in fact, it is this read). If we take a look at the BLAT alignment, we can see that the alignment is perfect, and you can define, at base-pair resolution, where the breakpoints are located, precisely at chr1:72,766,323 and at chr1:72,811,839. You could now design a pair of primers to try to validate this finding by an orthogonal technique such as Sanger sequencing.\n",
    "```console\n",
    "TGGATAAATA TGTGAGGGAG ACAATATAAA ACAAAAGTTC ACATGGAACA  72766229\n",
    "TACATTAATA AAATACTGCA GAAGCAAAAA AAAGTGAATC ATTCAAAAAT  72766279\n",
    "GAAAAGGGTA CAAGACAAAA ATAAAAGTTA AAGAAAAACC AAATgcaata  72766329\n",
    "gtagaattaa aattcatatt ctaaatatta aaacataaaa acatacaagt  72766379\n",
    "agtattgaca tttcagcaaa ttggataagt gatatggagg aaaaacatga  72766429\n",
    "agaaatatcc agatacctga gggaaaaaaa tgtgatgtat aataagtatt  72766479\n",
    "...\n",
    "ctatctgagg tcctacagta aaaagtggag tatggtaggc agaatactaa  72811729\n",
    "gatggcctgt caaggtttct cgaccatgtc attcattaaa acattaatct  72811779\n",
    "aggtgctgtt gtgaagagat tttgttgatt taaatgaagt cccagcttgg  72811829\n",
    "tagattttaa GACTGAGAGA TTATCTCAGT GGATTTGACT TAATCTCTTG  72811879\n",
    "AACCACTTTC TAAAAAgAGT TTTAgTTGGC TAGTTGCAGA AAAGAAAATC  72811929\n",
    "AGAGtTTCAA AGCTTTaagg ggattcaatg tgccactgat gacttgtaga  72811979\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In summary, this is the general procedure to detect CNVs and other structural variants from NGS data. However, there are specialized programs to detect CNVs by taking into consideration not only the number of reads, but the GC content and mappability of the region, improving the ability to detect CNVs. Similarly, the detection of SVs based on the combination of abnormally paired reads (discordant reads) as well as splitreads, allows the detection of different types of SVs (including duplications, inversions, translocations,...)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**You can now try it yourself. You can use the same case, but try on chromosome 12, between 10,000,000 and 11,000,000.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
